File: frontend/src/types/crowdfunding.ts
export type RepaymentType = 'in_fine' | 'amortizable'
export type ProjectStatus = 'funding' | 'active' | 'completed' | 'delayed' | 'defaulted'

/** Labels FR partagés (badges de statut) — source unique pour toutes les pages crowdfunding. */
export const STATUS_LABELS: Record<ProjectStatus, string> = {
  funding: 'En cours de levée',
  active: 'Actif',
  completed: 'Terminé',
  delayed: 'En retard',
  defaulted: 'Défaut',
}

/** Classes de couleur partagées pour les badges de statut. */
export const STATUS_COLORS: Record<ProjectStatus, string> = {
  funding: 'bg-warning/10 text-warning',
  active: 'bg-gain/10 text-gain',
  completed: 'bg-accent/10 text-accent',
  delayed: 'bg-warning/10 text-warning',
  defaulted: 'bg-loss/10 text-loss',
}

export interface ProjectDocument {
  id: string
  project_id: string
  file_name: string
  file_size: number
  audit_id: string | null
  created_at: string
}

export type PaymentType = 'interest' | 'capital' | 'both'

export interface CrowdfundingRepayment {
  id: string
  project_id: string
  payment_date: string
  amount: number
  payment_type: PaymentType
  interest_amount: number | null
  capital_amount: number | null
  tax_amount: number | null
  notes: string | null
  created_at: string
}

export type ScheduleStatus = 'paid' | 'pending' | 'overdue'
export type InterestFrequency = 'at_maturity' | 'monthly' | 'quarterly' | 'semi_annual' | 'annual'

export interface PaymentScheduleEntry {
  id: string
  project_id: string
  due_date: string
  expected_capital: number
  expected_interest: number
  is_completed: boolean
  completed_at: string | null
  repayment_id: string | null
  status: ScheduleStatus
}

export interface CrowdfundingProject {
  id: string
  asset_id: string
  platform: string
  project_name: string | null
  description: string | null
  project_url: string | null
  invested_amount: number
  annual_rate: number
  duration_months: number
  repayment_type: RepaymentType
  interest_frequency: InterestFrequency | null
  tax_rate: number
  delay_months: number
  start_date: string | null
  estimated_end_date: string | null
  actual_end_date: string | null
  status: ProjectStatus
  total_received: number
  created_at: string
  updated_at: string
  /** Intérêts projetés NETS de flat tax. */
  projected_total_interest: number | null
  /** Intérêts projetés BRUTS de fiscalité (invested × rate × months/12) — base homogène face à interest_earned (brut). */
  projected_interest_gross: number | null
  /** Intérêts réellement encaissés à date (bruts, avant prélèvements). */
  interest_earned: number | null
  progress_percent: number | null
  documents: ProjectDocument[]
  repayments: CrowdfundingRepayment[]
  schedule: PaymentScheduleEntry[]
}

export interface CrowdfundingCreateData {
  portfolio_id: string
  platform: string
  project_name: string
  description?: string
  project_url?: string
  invested_amount: number
  annual_rate: number
  duration_months: number
  repayment_type: RepaymentType
  interest_frequency?: InterestFrequency
  tax_rate?: number
  delay_months?: number
  start_date?: string
  estimated_end_date?: string
  status?: ProjectStatus
}

export interface CrowdfundingUpdateData {
  project_name?: string
  description?: string
  project_url?: string
  platform?: string
  annual_rate?: number
  duration_months?: number
  repayment_type?: RepaymentType
  interest_frequency?: InterestFrequency
  tax_rate?: number
  delay_months?: number
  start_date?: string
  estimated_end_date?: string
  actual_end_date?: string
  status?: ProjectStatus
  total_received?: number
}

export interface CrowdfundingDashboard {
  total_invested: number
  total_received: number
  /** Intérêts réellement encaissés (seul vrai P&L — le capital remboursé n'est pas un gain). */
  total_interest_received: number
  /** Capital déjà remboursé (retour de principal, pas un gain). */
  total_capital_repaid: number
  /** Capital restant dû des projets sains — la valeur réelle de la poche. */
  capital_outstanding: number
  /** Principal exposé sur projets en défaut (provisionné en perte). */
  defaulted_outstanding: number
  projected_annual_interest: number
  weighted_average_rate: number
  active_count: number
  completed_count: number
  delayed_count: number
  defaulted_count: number
  funding_count: number
  next_maturity: string | null
  /** Montants investis par plateforme (coût historique). */
  platform_breakdown: Record<string, number>
  /** Exposition par plateforme au capital restant dû (projets en défaut exclus). Optionnel pour compat backend. */
  platform_breakdown_outstanding?: Record<string, number>
  projects: CrowdfundingProject[]
}

export interface GuaranteeInfo {
  type: string
  rank: string | null
  description: string
  strength: 'forte' | 'moyenne' | 'faible'
}

export interface InvestmentSimulation {
  investment_amount: number
  duration_months: number
  tri_percent: number
  gross_interest: number
  tax_amount: number
  net_interest: number
  monthly_gross_return: number
  total_at_end: number
  roi_net_percent: number
}

export interface ProjectAudit {
  id: string
  project_id: string | null
  file_names: string[]
  document_type: string | null
  project_name: string | null
  operator: string | null
  location: string | null
  tri: number | null
  duration_min: number | null
  duration_max: number | null
  collection_amount: number | null
  margin_percent: number | null
  ltv: number | null
  ltc: number | null
  pre_sales_percent: number | null
  equity_contribution: number | null
  guarantees: GuaranteeInfo[]
  admin_status: string | null
  score_operator: number | null
  score_location: number | null
  score_guarantees: number | null
  score_risk_return: number | null
  score_admin: number | null
  risk_score: number | null
  points_forts: string[]
  points_vigilance: string[]
  red_flags: string[]
  verdict: 'INVESTIR' | 'VIGILANCE' | 'NE_PAS_INVESTIR'
  suggested_investment: number | null
  diversification_impact: string | null
  correlation_score: number | null
  portfolio_concentration: Record<string, number> | null
  investment_simulation: InvestmentSimulation | null
  created_at: string
}

export interface StressTestCashflow {
  date: string
  capital: number
  interest: number
  total: number
  is_delayed: boolean
}

export interface StressTestResult {
  project_id: string
  delay_months: number
  base_irr: number | null
  stressed_irr: number | null
  irr_delta: number | null
  cashflows: StressTestCashflow[]
}

export interface CrowdfundingPerformanceItem {
  id: string
  project_name: string | null
  platform: string
  status: ProjectStatus
  invested_amount: number
  annual_rate: number
  duration_months: number
  repayment_type: RepaymentType
  /** Intérêts projetés NETS de flat tax. */
  projected_total_interest: number
  /** Intérêts projetés BRUTS de fiscalité — à comparer à interest_earned (brut). */
  projected_interest_gross: number
  /** Total reçu (brut, capital + intérêts) — ne pas comparer aux intérêts projetés. */
  total_received: number
  /** Intérêts réellement encaissés (bruts). */
  interest_earned: number
  elapsed_months: number
  progress_percent: number
  on_track: boolean
  start_date: string | null
  estimated_end_date: string | null
}

File: backend/app/schemas/crowdfunding.py
"""Pydantic schemas for crowdfunding projects."""

from datetime import date, datetime
from decimal import Decimal
from typing import Optional
from uuid import UUID

from pydantic import BaseModel, Field, computed_field, field_validator

from app.models.crowdfunding_project import ProjectStatus, RepaymentType
from app.models.crowdfunding_repayment import PaymentType
from app.schemas._money import Money

# ---------- Request schemas ----------


class CrowdfundingProjectCreate(BaseModel):
    portfolio_id: Optional[UUID] = None
    platform: str = Field(..., max_length=100)
    project_name: str = Field(..., max_length=300)
    description: Optional[str] = None
    project_url: Optional[str] = Field(None, max_length=500)
    invested_amount: Money = Field(..., ge=0)
    annual_rate: Money = Field(..., ge=0, le=100)
    duration_months: int = Field(..., ge=1, le=360)
    repayment_type: RepaymentType = RepaymentType.IN_FINE
    start_date: Optional[date] = None
    estimated_end_date: Optional[date] = None
    interest_frequency: Optional[str] = Field("at_maturity", max_length=20)
    tax_rate: Money = Field(Decimal("30.00"), ge=0, le=100)
    delay_months: int = Field(0, ge=0, le=120)
    status: ProjectStatus = ProjectStatus.ACTIVE


class CrowdfundingProjectUpdate(BaseModel):
    project_name: Optional[str] = Field(None, max_length=300)
    description: Optional[str] = None
    project_url: Optional[str] = Field(None, max_length=500)
    platform: Optional[str] = Field(None, max_length=100)
    invested_amount: Optional[Money] = Field(None, gt=0)
    annual_rate: Optional[Money] = Field(None, ge=0, le=100)
    duration_months: Optional[int] = Field(None, ge=1, le=360)
    repayment_type: Optional[RepaymentType] = None
    start_date: Optional[date] = None
    estimated_end_date: Optional[date] = None
    actual_end_date: Optional[date] = None
    status: Optional[ProjectStatus] = None
    interest_frequency: Optional[str] = Field(None, max_length=20)
    tax_rate: Optional[Money] = Field(None, ge=0, le=100)
    delay_months: Optional[int] = Field(None, ge=0, le=120)
    total_received: Optional[Money] = Field(None, ge=0)


# ---------- Repayment schemas ----------


class RepaymentCreate(BaseModel):
    payment_date: date
    amount: Money = Field(..., gt=0)
    payment_type: PaymentType
    interest_amount: Optional[Money] = Field(None, ge=0)
    capital_amount: Optional[Money] = Field(None, ge=0)
    tax_amount: Optional[Money] = Field(None, ge=0)
    notes: Optional[str] = None


class RepaymentResponse(BaseModel):
    id: UUID
    project_id: UUID
    payment_date: date
    amount: Money
    payment_type: PaymentType
    interest_amount: Optional[Money] = None
    capital_amount: Optional[Money] = None
    tax_amount: Optional[Money] = None
    notes: Optional[str] = None
    created_at: datetime

    class Config:
        from_attributes = True


# ---------- Response schemas ----------


class ProjectDocumentResponse(BaseModel):
    id: UUID
    project_id: UUID
    file_name: str
    file_size: int
    audit_id: Optional[UUID] = None
    created_at: datetime

    class Config:
        from_attributes = True


class PaymentScheduleEntryResponse(BaseModel):
    id: UUID
    project_id: UUID
    due_date: date
    expected_capital: Money
    expected_interest: Money
    is_completed: bool
    completed_at: Optional[datetime] = None
    repayment_id: Optional[UUID] = None
    status: str = "pending"  # "paid" | "pending" | "overdue"

    class Config:
        from_attributes = True


class CrowdfundingProjectResponse(BaseModel):
    id: UUID
    asset_id: UUID
    platform: str
    project_name: Optional[str] = None
    description: Optional[str] = None
    project_url: Optional[str] = None
    invested_amount: Money
    annual_rate: Money
    duration_months: int
    repayment_type: RepaymentType
    interest_frequency: Optional[str] = "at_maturity"
    tax_rate: Money = Decimal("30.00")
    delay_months: int = 0
    start_date: Optional[date] = None
    estimated_end_date: Optional[date] = None
    actual_end_date: Optional[date] = None
    status: ProjectStatus
    total_received: Money
    created_at: datetime
    updated_at: datetime

    # Computed
    projected_total_interest: Optional[float] = None
    # Intérêts projetés BRUTS de fiscalité (invested × rate × months/12) —
    # permet des ratios homogènes brut/brut face à interest_earned (brut).
    projected_interest_gross: Optional[float] = None
    interest_earned: Optional[float] = None
    progress_percent: Optional[float] = None
    documents: list[ProjectDocumentResponse] = []
    repayments: list[RepaymentResponse] = []
    schedule: list[PaymentScheduleEntryResponse] = []

    class Config:
        from_attributes = True


class CrowdfundingDashboardResponse(BaseModel):
    total_invested: float
    total_received: float
    # Décomposition comptable : les intérêts sont du P&L, le capital remboursé
    # est un retour de principal ; le CRD (capital restant dû) est la valeur
    # de la poche ; le principal des projets en défaut est provisionné à part.
    total_interest_received: float = 0.0
    total_capital_repaid: float = 0.0
    capital_outstanding: float = 0.0
    defaulted_outstanding: float = 0.0
    # Bonus de parrainage / plateforme encaissés (hors intérêts et hors capital).
    total_referral: float = 0.0
    projected_annual_interest: float
    weighted_average_rate: float
    active_count: int
    completed_count: int
    delayed_count: int
    defaulted_count: int
    funding_count: int
    next_maturity: Optional[date] = None
    platform_breakdown: dict[str, float]
    # Exposition par plateforme au capital restant dû (projets en défaut
    # exclus) — un projet remboursé à 80 % ne pèse plus que 20 %.
    platform_breakdown_outstanding: dict[str, float] = Field(default_factory=dict)
    projects: list[CrowdfundingProjectResponse]


# ---------- Cashflow schedule (échéancier consolidé) ----------


class CashflowProjectAmount(BaseModel):
    """Contribution d'un projet au cash-flow d'un mois (montant BRUT de fiscalité)."""

    name: str
    amount: float


class CashflowMonthResponse(BaseModel):
    """Cash-flows attendus sur un mois, agrégés sur tous les projets non-défaut.

    Somme des échéances contractuelles NON complétées (capital + intérêts).
    Les montants sont BRUTS de fiscalité — les échéanciers sont contractuels,
    avant prélèvements.
    """

    month: str  # "YYYY-MM"
    expected_capital: float = 0.0
    expected_interest: float = 0.0
    total: float = 0.0
    projects: list[CashflowProjectAmount] = Field(default_factory=list)


# ---------- Rapport fiscal annuel (réconciliation IFU) ----------


class TaxReportPlatform(BaseModel):
    """Agrégat fiscal annuel d'une plateforme.

    Les intérêts de crowdfunding immobilier (obligations) sont des revenus de
    capitaux mobiliers soumis au PFU 30 % (12,8 % IR + 17,2 % PS), prélevé À LA
    SOURCE par la plateforme (le ``tax_amount`` saisi sur chaque versement) et
    déclaré case 2TR/2BH via l'IFU (formulaire 2561) qu'elle envoie. Ce rapport
    sert à réconcilier ces IFU avec les versements saisis dans l'app.

    ``withholding_gap`` : True quand la retenue saisie diverge de plus de 1 €
    du PFU théorique (``gross_interest × 0.30``) — soit une dispense d'acompte
    (12,8 % IR non prélevé sur demande, revenu fiscal de référence sous les
    seuils), soit des versements enregistrés sans le split fiscal renseigné.
    """

    platform: str
    # Intérêts bruts encaissés dans l'année (avant prélèvements)
    gross_interest: float
    # Retenue à la source saisie sur les versements (somme des tax_amount)
    tax_withheld: float
    # Net perçu = brut − retenues
    net_interest: float
    # PFU théorique : gross_interest × 30 %
    theoretical_pfu: float
    withholding_gap: bool
    nb_payments: int


class CrowdfundingTaxReportResponse(BaseModel):
    """Rapport fiscal annuel agrégé par plateforme + totaux."""

    year: int
    platforms: list[TaxReportPlatform]
    total_gross_interest: float
    total_tax_withheld: float
    total_net_interest: float
    total_theoretical_pfu: float
    nb_payments: int


# ---------- Audit Lab schemas ----------


class GuaranteeInfo(BaseModel):
    type: str
    rank: Optional[str] = None
    description: str
    strength: str  # "forte", "moyenne", "faible"


class ProjectAuditResponse(BaseModel):
    id: UUID
    project_id: Optional[UUID] = None
    file_names: list[str]
    document_type: Optional[str] = None
    project_name: Optional[str] = None
    operator: Optional[str] = None
    location: Optional[str] = None
    tri: Optional[float] = None
    duration_min: Optional[int] = None
    duration_max: Optional[int] = None
    collection_amount: Optional[float] = None
    margin_percent: Optional[float] = None
    ltv: Optional[float] = None
    ltc: Optional[float] = None
    pre_sales_percent: Optional[float] = None
    equity_contribution: Optional[float] = None
    guarantees: list[GuaranteeInfo] = []
    admin_status: Optional[str] = None
    score_operator: Optional[int] = None
    score_location: Optional[int] = None
    score_guarantees: Optional[int] = None
    score_risk_return: Optional[int] = None
    score_admin: Optional[int] = None
    risk_score: Optional[int] = None
    points_forts: list[str] = []
    points_vigilance: list[str] = []
    red_flags: list[str] = []
    verdict: str = "VIGILANCE"
    suggested_investment: Optional[float] = None
    diversification_impact: Optional[str] = None
    correlation_score: Optional[float] = None
    portfolio_concentration: Optional[dict] = None
    created_at: datetime

    @field_validator("guarantees", "points_forts", "points_vigilance", "red_flags", mode="before")
    @classmethod
    def none_to_list(cls, v):  # noqa: N805
        return v if v is not None else []

    @field_validator("verdict", mode="before")
    @classmethod
    def none_to_default_verdict(cls, v):  # noqa: N805
        return v if v is not None else "VIGILANCE"

    @computed_field  # type: ignore[misc]
    @property
    def investment_simulation(self) -> Optional[dict]:
        """Compute investment simulation from audit data."""
        if self.suggested_investment is None or self.tri is None:
            return None

        invest = self.suggested_investment
        tri_pct = self.tri / 100.0
        dur_min = self.duration_min or 12
        dur_max = self.duration_max or dur_min
        dur_avg = (dur_min + dur_max) / 2.0
        tax_rate = 0.30  # Flat tax France

        gross_interest = round(invest * tri_pct * (dur_avg / 12.0), 2)
        net_interest = round(gross_interest * (1 - tax_rate), 2)
        monthly_gross = round(gross_interest / dur_avg, 2) if dur_avg > 0 else 0
        total_at_end = round(invest + net_interest, 2)

        return {
            "investment_amount": invest,
            "duration_months": round(dur_avg),
            "tri_percent": self.tri,
            "gross_interest": gross_interest,
            "tax_amount": round(gross_interest * tax_rate, 2),
            "net_interest": net_interest,
            "monthly_gross_return": monthly_gross,
            "total_at_end": total_at_end,
            "roi_net_percent": round((net_interest / invest) * 100, 2) if invest > 0 else 0,
        }

    class Config:
        from_attributes = True


# ---------- Stress Test schemas ----------


class StressTestCashflowSchema(BaseModel):
    date: date
    capital: float
    interest: float
    total: float
    is_delayed: bool


class StressTestResponse(BaseModel):
    project_id: UUID
    delay_months: int
    base_irr: Optional[float] = None
    stressed_irr: Optional[float] = None
    irr_delta: Optional[float] = None
    cashflows: list[StressTestCashflowSchema]
